grpcclient/grpcclient_test.go

```diff
@@ -0,0 +1,121 @@
+		return false
+	}
+
+	for i, arg := range o.callOpts {
+		if maxRecv, ok := arg.(grpc.MaxRecvMsgSizeCallOption); ok {
+			if maxRecv != other.callOpts[i].(grpc.MaxRecvMsgSizeCallOption) {
+				return false
+			}
+			continue
+		}
+		if maxSend, ok := arg.(grpc.MaxSendMsgSizeCallOption); ok {
+			if maxSend != other.callOpts[i].(grpc.MaxSendMsgSizeCallOption) {
+				return false
+			}
+			continue
+		}
+	}
+
+	hasUnaryInterceptor := o.unaryClientInterceptor != nil
+	otherHasUnaryInterceptor := other.unaryClientInterceptor != nil
+	hasStreamInterceptor := o.streamClientInterceptor != nil
+	otherHasStreamInterceptor := other.streamClientInterceptor != nil
+
+	return o.isInsecure == other.isInsecure && hasUnaryInterceptor == otherHasUnaryInterceptor &&
+		hasStreamInterceptor == otherHasStreamInterceptor && o.keepaliveParams == other.keepaliveParams
+}
+
+func TestConfig(t *testing.T) {
+	origWithDefaultCallOptions := withDefaultCallOptions
+	origWithUnaryInterceptor := withUnaryInterceptor
+	origWithStreamInterceptor := withStreamInterceptor
+	origWithKeepaliveParams := withKeepaliveParams
+	origWithInsecure := tls.WithInsecure
+	t.Cleanup(func() {
+		withDefaultCallOptions = origWithDefaultCallOptions
+		withUnaryInterceptor = origWithUnaryInterceptor
+		withStreamInterceptor = origWithStreamInterceptor
+		withKeepaliveParams = origWithKeepaliveParams
+		tls.WithInsecure = origWithInsecure
+	})
+
+	withDefaultCallOptions = func(cos ...grpc.CallOption) grpc.DialOption {
+		t.Log("Received call options", "options", cos)
+		return fakeDialOption{callOpts: cos}
+	}
+	withUnaryInterceptor = func(f grpc.UnaryClientInterceptor) grpc.DialOption {
+		t.Log("Received unary client interceptor", f)
+		return fakeDialOption{unaryClientInterceptor: f}
+	}
+	withStreamInterceptor = func(f grpc.StreamClientInterceptor) grpc.DialOption {
+		t.Log("Received stream client interceptor", f)
+		return fakeDialOption{streamClientInterceptor: f}
+	}
+	withKeepaliveParams = func(kp keepalive.ClientParameters) grpc.DialOption {
+		t.Log("Received keepalive params", kp)
+		return fakeDialOption{
+			keepaliveParams: kp,
+		}
+	}
+	tls.WithInsecure = func() grpc.DialOption {
+		return fakeDialOption{isInsecure: true}
+	}
+
+	cfg := Config{
+		KeepaliveTime:    10,
+		KeepaliveTimeout: 20,
+	}
+	expOpts := []grpc.DialOption{
+		fakeDialOption{isInsecure: true},
+		fakeDialOption{callOpts: []grpc.CallOption{
+			grpc.MaxCallRecvMsgSize(0),
+			grpc.MaxCallSendMsgSize(0),
+		}},
+		fakeDialOption{
+			unaryClientInterceptor: middleware.ChainUnaryClient(),
+		},
+		fakeDialOption{
+			streamClientInterceptor: middleware.ChainStreamClient(),
+		},
+		fakeDialOption{
+			keepaliveParams: keepalive.ClientParameters{
+				Time:                10 * time.Second,
+				Timeout:             20 * time.Second,
+				PermitWithoutStream: true,
+			},
+		},
+	}
+
+	opts, err := cfg.DialOption(nil, nil)
+	require.NoError(t, err)
+
+	assert.Empty(t, cmp.Diff(expOpts, opts))
```

Comment:
assert.Equals will already print expected and real values if the check fails. Perhaps not as nice, but doesn’t require another dependency.

Reply:
@pstibrany I can try `assert.Equals` again, but it wasn't working for me initially due to being too naïve (this is the benefit of go-cmp, that it's a more advanced comparer).

Reply:
I won't block the PR on that, but will point out that expected outcome is no difference. Bringing in extra library that is only useful in case of test failure seems unnecessary.

Reply:
Oh, I misunderstood the purpose then. I thought it's rich diff you're after, not just equality check.

Reply:
@pstibrany yes exactly :) Rich diff is nice, but I'm only bringing it in here because it actually is needed.